pennylane/devices/default_qubit_numba.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains an numba implementation of the :class:`~.DefaultQubit`
+reference plugin.
+"""
+from pennylane.operation import DiagonalOperation
+from pennylane.devices import DefaultQubit
+from pennylane.devices import numba_ops
+import numpy as np
+
+
+class DefaultQubitNumba(DefaultQubit):
+    """Simulator plugin based on ``"default.qubit"``, written using numba.
+
+    **Short name:** ``default.qubit.numba``
+
+    This device provides a pure-state qubit simulator written using numba. As a result, it
+    supports classical backpropagation as a means to compute the gradient. This can be faster than
+    the parameter-shift rule for analytic quantum gradients when the number of parameters to be
+    optimized is large.
+
+    To use this device, you will need to install numba:
+
+    .. code-block:: console
+
+        pip install numba
+
+    Using this method, the created QNode is a 'white-box', and is
+    tightly integrated with your numba computation:
+
+    >>> dev = qml.device("default.qubit.numba", wires=1)
+    >>> @qml.qnode(dev)
+    ... def circuit(x):
+    ...     qml.RX(x[1], wires=0)
+    ...     qml.Rot(x[0], x[1], x[2], wires=0)
+    ...     return qml.expval(qml.PauliZ(0))
+    >>> weights = np.array([0.2, 0.5, 0.1])
+    >>> grad_fn = qml.grad(circuit)
+    >>> print(grad_fn(weights))
+    array([-2.25267173e-01 -1.00864546e+00  5.55111512e-17])
+
+
+    Args:
+        wires (int): the number of wires to initialize the device with
+        shots (None, int): How many times the circuit should be evaluated (or sampled) to estimate
+            the expectation values. Defaults to ``None`` if not specified, which means that the device
+            returns analytical results.
+        analytic (bool): Indicates if the device should calculate expectations
+            and variances analytically. In non-analytic mode, the ``diff_method="backprop"``
+            QNode differentiation method is not supported and it is recommended to consider
+            switching device to ``default.qubit`` and using ``diff_method="parameter-shift"``.
+    """
+
+    name = "Default qubit (Numba) PennyLane plugin"
+    short_name = "default.qubit.numba"
+
+    parametric_ops = {
+        "PhaseShift": numba_ops.PhaseShift,
```

Comment:
How about we instead use the class object itself for the mapping dictionary:

I think this way is mildly more efficient, but maybe not by enough that it matters.
```
parametric_ops = {
    qml.PhaseShift: numba_ops.PhaseShift,
}
f = parametric_ops[op.__class__]
```